src/server/undolr.c

```diff
@@ -0,0 +1,213 @@
+ * @brief
+ *	The Undo Recording API calls to integrate with PBSPro.
+ *	It allows our daemons to create an Undo Recording of itself running, which can
+ *	then be opened using the Undo Debugger/Player (UndoDB).
+ * 
+ */
+
+#include <signal.h>
+#include <string.h>
+#include <unistd.h>
+#include <assert.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <time.h>
+#include <errno.h>
+#include <pbs_internal.h>
+#include "log.h"
+#include "undolr.h"
+
+int sigusr1_flag = 0;
+static int recording = 0;
+static char recording_file [MAXPATHLEN + 1] = {0};
+
+/**
+ * @brief
+ *	catch_sigusr1() - the signal handler for  SIGUSR1.
+ *	Set a flag for the main loop to know that a sigusr1 processes
+ *
+ * @param[in]	sig	- not used in fun.
+ *
+ * @return	void
+ */
+void
+catch_sigusr1(int sig)
+{
+    /*TODO: remove below logs */
+    sprintf(log_buffer, "%s: caught signal %d", __func__, sig);
+	log_event(PBSEVENT_SYSTEM | PBSEVENT_FORCE, PBS_EVENTCLASS_SERVER,
+		LOG_NOTICE, msg_daemonname, log_buffer);
+	sigusr1_flag = 1;
+}
+
+/**
+ * @brief
+ *	mk_recording_path - make the recording name and path used by deamons 
+ *	based on the date and time: <deamon_name>_yyyymmddHHMM.undo
+ *
+ * @param[in] fpath - buffer to hold the live recording file path
+ * @param[in] maxlen - max size of buffer
+ *
+ * @return  0 - Success
+ *          1 - Failure
+ *
+ */
+static void mk_recording_path(char * fpath, int maxlen) 
+{
+	struct tm ltm;
+	struct tm *ptm;
+	time_t time_now;
+
+	if (pbs_loadconf(1) == 0) {
+		sprintf(log_buffer, "%s: Could not load pbs configuration", __func__);
+		log_event(PBSEVENT_ERROR, PBS_EVENTCLASS_SERVER, LOG_ALERT,
+			msg_daemonname, log_buffer);
+	}
+
+	time_now = time(NULL);
+	ptm = localtime_r(&time_now, &ltm);
+
+	if (pbs_conf.pbs_lr_save_path)
+		(void)snprintf(fpath, maxlen,
+			"%s/%s_%04d%02d%02d%02d%02d.undo",
+			pbs_conf.pbs_lr_save_path, msg_daemonname, ptm->tm_year+1900, ptm->tm_mon+1,
+			ptm->tm_mday, ptm->tm_hour,ptm->tm_min);
+	else /* default path */
+		(void)snprintf(fpath, maxlen,
+			"%s/%s/%s_%04d%02d%02d%02d%02d.undo",
+			pbs_conf.pbs_home_path, "spool", msg_daemonname, ptm->tm_year+1900, ptm->tm_mon+1,
+			ptm->tm_mday, ptm->tm_hour,ptm->tm_min);
+}
+
+/**
+ * @brief 
+ *	undolr - call respective Undo Live Recorder APIs
+ *	to start and stop the recordings.
+ * @return  void 
+ */
+void undolr()
+{
+		int e = 0;
+		undolr_error_t  err = 0;
+		undolr_recording_context_t lr_ctx;
+
+	if (!recording)
+	{
+		mk_recording_path(recording_file, MAXPATHLEN);
+		sprintf(log_buffer,
+				"Undo live recording started, will save to %s", recording_file);
+		log_event(PBSEVENT_ADMIN | PBSEVENT_FORCE, 
+				PBS_EVENTCLASS_SERVER, LOG_DEBUG, msg_daemonname, log_buffer);
+
+		/**
+		 * Undo API call to start recording.
+		 * Attaches Live Recorder to the process, and starts recording it.
+		 */
+		e = undolr_start(&err);
+		if (e)
+		{
+			sprintf(log_buffer,
+				"undolr_recording_start() failed: error=%i errno=%i\n", e, errno);
```

Comment:
we do not have undolr_recording_start(), we can simply say - unable to start undo recording along with error and errno.